### PR TITLE
Selectively log warning or error depending on how much time elapsed since bootstrapping finished

### PR DESCRIPTION
**Pull request by yacovm**
## Why this should be merged

This commit adjusts the log level of the unexpected build block failure according to how much time passed since bootstrapping finished.

If less than 5 minutes elapsed then it's a warning. Else it's an error.

## How this works

Sets the bootstrapping time when we finished bootstrapping and according to the elapsed time, we choose to log in error or warning.

## How this was tested

UT

### CODE REVIEW

File: vms/proposervm/block.go [hunk 1]
```diff
@@ -23,7 +23,8 @@ import (
 
 const (
 	// allowable block issuance in the future
-	maxSkew = 10 * time.Second
+	maxSkew                         = 10 * time.Second
+	bootstrappingWarningGracePeriod = 5 * time.Minute
```
Comment (Elvis339) [inline vms/proposervm/block.go:28]: How did you get this number?

Comment (maru-ava) [inline vms/proposervm/block.go:28]: (No action required) Would it make sense for this grace period to be configurable?

Comment (yacovm) [reply vms/proposervm/block.go:28]: chosen by a fair 🎲 roll.

**Comment by DracoLi** [comment]
i actually run into this error as well from our antithesis runs. I think the part that should be a warn instead of error is specifically the `validator set height not yet finalized` errors that can happen in `ExpectedProposer`, and not the other errors (something like [this](https://github.com/ava-labs/avalanchego/commit/693e123f6af7fee942652ce0149d11d34807737c)). I think that is better than just changing the log to warn for all "failed to calculate expected proposer" errors during boostrap time frame.

File: vms/proposervm/block.go [hunk 2]
```diff
@@ -514,3 +515,11 @@ func (p *postForkCommonComponents) shouldBuildSignedBlockPreDurango(
 	)
 	return false, fmt.Errorf("%w: delay %s < minDelay %s", errProposerWindowNotStarted, delay, minDelay)
 }
+
+func (p *postForkCommonComponents) logWarnOrError() func(msg string, fields ...zap.Field) {
```
Comment (maru-ava) [inline vms/proposervm/block.go:528]: (No action required) Maybe refactor into a function that can be trivially tested? 90+ lines of test code for a 7 line function is not ideal.

```golang
  func logLevelSinceBootstrapping(now, finished time.Time) zapcore.Level {
        if finished.IsZero() || now.Sub(finished) < bootstrappingWarningGracePeriod {
                return zap.WarnLevel
        }
        return zap.ErrorLevel
  }

  func (p *postForkCommonComponents) logWarnOrError() func(msg string, fields ...zap.Field) {
        switch logLevelSinceBootstrapping(p.vm.Clock.Time(), p.vm.finishedBootstrappingAt) {
        case zap.WarnLevel:
                return p.vm.ctx.Log.Warn
        default:
                return p.vm.ctx.Log.Error
        }
  }
```

File: vms/proposervm/vm.go [hunk 3]
```diff
@@ -316,6 +317,11 @@ func (vm *VM) SetState(ctx context.Context, newState snow.State) error {
 	}
 
 	oldState := vm.consensusState
+
+	if newState == snow.NormalOp && oldState != snow.NormalOp {
```
Comment (maru-ava) [inline vms/proposervm/vm.go:321]: (No action required) Maybe test that the time is set separately from choosing the log level?

**Comment by yacovm** [comment]
> i actually run into this error as well from our antithesis runs. I think the part that should be a warn instead of error is specifically the `validator set height not yet finalized` errors that can happen in `ExpectedProposer`, and not the other errors (something like [this](https://github.com/ava-labs/avalanchego/commit/693e123f6af7fee942652ce0149d11d34807737c)). I think that is better than just changing the log to warn for all "failed to calculate expected proposer" errors during boostrap time frame.

I don't think it's correct to always assume this is a warning, because if this happens after bootstrapping this is a bug, because it means we're trying to build a block on top of a block with a P-chain height that we don't have, but we should've had this P-chain height.

File: vms/proposervm/vm.go (hunk 3 above)
Comment (yacovm) [reply vms/proposervm/vm.go:321]: Not sure I follow, can you elaborate what you mean?

File: vms/proposervm/block.go (hunk 2 above)
Comment (yacovm) [reply vms/proposervm/block.go:528]: The 90+ LOC is not for the function, it's for testing the function is properly invoked in the VM.

File: vms/proposervm/block.go (hunk 1 above)
Comment (yacovm) [reply vms/proposervm/block.go:28]: I don't think so, because all this determines is whether we log in error or warning. From observations, this shouldn't happen within a timeframe of 5 minutes as this usually happens right after bootstrapping

File: vms/proposervm/block.go (hunk 2 above)
Comment (maru-ava) [reply vms/proposervm/block.go:528]: I see two changes in this PR that merit unit testing:
 - recording the time that bootstrap completes
 - using the time of bootstrap completion to determine whether an error or a warning is logged
 
I think these are separable concerns that can be trivially validated independently without worrying about whether the function is called by the VM. We can tell by observation that these are called, the logic is the part that needs validation. 

Is there value in testing the entire code path? Sure, but only if it's not too expensive. Testing always involves tradeoffs. The complexity presented in this PR suggests to me that it is not worth it in this instance. I think it would instead be preferable to avoid the cost to the implementor, reviewers, and future maintainers by pursuing a simpler strategy focused on the 2 separable concerns.

File: vms/proposervm/vm.go (hunk 3 above)
Comment (maru-ava) [reply vms/proposervm/vm.go:321]: Elaborated in the other comment.

File: vms/proposervm/block.go (hunk 2 above)
Comment (yacovm) [reply vms/proposervm/block.go:528]: I see your point, but I think that testing that the time of bootstrap completion was set correctly is wrong because it's an internal implementation detail of the VM. 

Besides, we need to trigger `logWarnOrError()` somehow anyway, and that function depends on the time we finished bootstrapping, and setting it manually in the test is wrong because that's not the natural flow of the component.

Comment (maru-ava) [reply vms/proposervm/block.go:528]: I don't think these are mutually exclusive perspectives. So maybe validate the separable concerns independently from the golden-path of ensuring they tie together? I think the current test is way too gnarly for what is intended by this PR.

Comment (maru-ava) [reply vms/proposervm/block.go:528]: > I see your point, but I think that testing that the time of bootstrap completion was set correctly is wrong because it's an internal implementation detail of the VM.

Why would it be inappropriate to validate the internal implementation detail of the VM?

Comment (maru-ava) [reply vms/proposervm/block.go:528]: > Besides, we need to trigger logWarnOrError() somehow anyway, and that function depends on the time we finished bootstrapping, and setting it manually in the test is wrong because that's not the natural flow of the component.

Again, separable concerns. Conflating separable concerns is definitely a way to ensure testing complexity, and I think it would be preferable to try to minimize that complexity.

Comment (yacovm) [reply vms/proposervm/block.go:528]: > I don't think these are mutually exclusive perspectives. So maybe validate the separable concerns independently from the golden-path of ensuring they tie together? I think the current test is way too gnarly for what is intended by this PR.

Currently the test tests both concerns, as you pointed out.

If I validate each concern in the test separately then the test flow will not model the natural flow of the component. 

Setting an inner field (`finishedBootstrappingAt`) artificially in the test, makes the test artificial and unrealistic because internal fields need to be set via the natural flow, unless they are configurations.

Comment (maru-ava) [reply vms/proposervm/block.go:528]: I'm not sure how relevant the terms 'natural flow' and 'artificial' might be for describing testing strategy. To me, asking "What is the quality of test coverage for this feature?" is instead a question of the degree of coverage and the cost/complexity of that coverage.

I think of effective testing strategy as ensuring that coverage is applied at as low a level of abstraction as possible to both minimize cost/complexity and to ensure that a test failure is easily attributable to the offending code. One strategy for achieving this can be [Functional Core, Imperative Shell](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell). Functional code - with its absence of side-effects - allows for comprehensive coverage at minimum cost/complexity. Since corner-cases are tested at the functional level, the imperative shell can then be validated with golden-path testing.

That said, I don't think this PR warrants reworking the test. If you're interested, I'd be happy to pair on testing strategy sometime. I think the `Functional Core, Imperative Shell` approach can really change how you think about where to put coverage.
